p2p/ping.go

```diff
@@ -0,0 +1,158 @@
+package p2p
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/routing"
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+	ma "github.com/multiformats/go-multiaddr"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	pingInterval   = time.Second
+	pingProtectTag = "spacemesh-ping"
+)
+
+type pingStat struct {
+	numSuccess int
+	numFail    int
+}
+
+type Ping struct {
+	sync.Mutex
+	logger *zap.Logger
+	h      host.Host
+	peers  []peer.ID
+	pr     routing.PeerRouting
+	stats  map[peer.ID]*pingStat
+	cancel context.CancelFunc
+	eg     errgroup.Group
+}
+
+func NewPing(logger *zap.Logger, h host.Host, peers []peer.ID, pr routing.PeerRouting) *Ping {
+	p := &Ping{
+		logger: logger,
+		h:      h,
+		peers:  peers,
+		pr:     pr,
+		stats:  make(map[peer.ID]*pingStat),
+	}
+	return p
+}
+
+func (p *Ping) Start(ctx context.Context) {
+	if p.cancel != nil {
+		return
+	}
+	var startCtx context.Context
+	startCtx, p.cancel = context.WithCancel(context.Background())
+	for _, peer := range p.peers {
+		p.startForPeer(startCtx, peer)
+	}
+}
+
+func (p *Ping) Stop() {
+	if p.cancel == nil {
+		return
+	}
+	p.cancel()
+	p.cancel = nil
+	p.eg.Wait()
+}
+
+func (p *Ping) doPing(ctx context.Context, peerID peer.ID) (<-chan ping.Result, error) {
+	_, err := p.pr.FindPeer(ctx, peerID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find peer %s: %w", peerID, err)
+	}
+
+	return ping.Ping(ctx, p.h, peerID), nil
+}
+
+// runForPeer runs ping for the specific peer till an error occurs
+// or the context is canceled.
+func (p *Ping) runForPeer(ctx context.Context, peerID peer.ID, addrs []ma.Multiaddr) error {
+	pingCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	ch, err := p.doPing(pingCtx, peerID)
+	if err != nil {
+		return err
+	}
+
+	for r := range ch {
+		p.record(peerID, r.Error == nil)
+		if r.Error != nil {
+			return r.Error
+		} else {
+			p.logger.Info("PING succeeded",
+				zap.Stringer("peer", peerID),
+				zap.Any("addrs", addrs),
+				zap.Duration("rtt", r.RTT))
+		}
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(pingInterval):
+		}
+	}
+
+	return nil
+}
+
+func (p *Ping) startForPeer(ctx context.Context, peerID peer.ID) {
+	p.h.ConnManager().Protect(peerID, pingProtectTag)
+	p.eg.Go(func() error {
+		for {
+			var addrs []ma.Multiaddr
+			for _, c := range p.h.Network().ConnsToPeer(peerID) {
+				addrs = append(addrs, c.RemoteMultiaddr())
+			}
+			err := p.runForPeer(ctx, peerID, addrs)
+			if errors.Is(err, context.Canceled) {
+				return nil
+			}
+			p.logger.Error("PING failed",
+				zap.Stringer("peer", peerID),
+				zap.Any("addrs", addrs),
+				zap.Error(err))
```

Comment:
Is error the correct logging level here, i.e. is something wrong with the node if this line appears in the log?

Reply:
`Ping` is a network troubleshooting feature which is intended to be only used for diagnostic purposes, namely, to understand connectivity issues between nodes which must be able to talk to each other but somehow can't. When it is enabled, it's preferable to give ping messages higher visibility even if logging level is reduced, thus I use `Error` for ping errors and `Info` for ping successes.